### PR TITLE
haptic feedback support

### PR DESCRIPTION
**Pull request by pazos**
Improvement over #190 

I tested this on Moto G5S with great results. Probably needs to be tested in a device with a supported freescale e-ink screen and an haptic engine it something like that already exists.

### CODE REVIEW

File: assets/android.lua [hunk 1]
```diff
@@ -323,6 +323,12 @@ enum {
     ATIMEOUT_SYSTEM = 0,
 };
 
+enum {
+    AHAPTIC_LONG_PRESS = 0,
+    AHAPTIC_KEYBOARD_TAP = 3,
```
Comment (Frenzie) [inline assets/android.lua]: These are just the few I implemented as a kind of demo. It's probably simpler to just add the whole lot of them At the very `TEXT_HANDLE_MOVE` might potentially be useful in the future.

File: assets/android.lua [hunk 2]
```diff
@@ -1475,6 +1481,17 @@ local function run(android_app_state)
...
     end
 
+    android.hapticFeedback = function(type)
+        android.DEBUG("haptic feedback: "..type)
+        JNI:context(android.app.activity.vm, function(JNI)
+            JNI:callVoidMethod(
+                android.app.activity.clazz,
+                "performHapticFeedback",
+                "(I)V",
+                ffi.new('int32_t', type)
```
Comment (Frenzie) [inline assets/android.lua]: This is always correct and not platform specific?

File: assets/android.lua (hunk 1 above)
Comment (pazos) [reply assets/android.lua]: TEXT_HANDLE_MOVE was introduced in API 27. Most of the others were introduced in recent apis too.

I don't think using an unrecognized int will trigger an exception but I cannot test as the call is never triggered in the emulator and I currently don't have any working phone on any api below 28.

Shall I add the FLAG_IGNORE_GLOBAL_SETTING and FLAG_IGNORE_VIEW_SETTING?. These settings are useful to override (haptic) system settings.

File: assets/android.lua (hunk 2 above)
Comment (pazos) [reply assets/android.lua]: Yeah, it calls https://github.com/koreader/android-luajit-launcher/pull/192/files#diff-272d98f2e8dcd4377fdca0ca2a53178aR181.

That function gets a view and calls https://github.com/koreader/android-luajit-launcher/pull/192/files#diff-272d98f2e8dcd4377fdca0ca2a53178aR194.

Finally view.performHapticFeedback(constant) will run only if the device has an haptic feedback engine and the setting wasn't disabled for that specific view or globally.

**Comment by pazos** [comment]
> No comment on the other commits.

I will strip the last commit and submit it on another PR. I was testing it :p

File: assets/android.lua (hunk 2 above)
Comment (Frenzie) [reply assets/android.lua]: I mean like 32-bit vs 64-bit. ;-)

Comment (pazos) [reply assets/android.lua]: oh! Based on this we're good: https://stackoverflow.com/a/15005226

File: assets/android.lua (hunk 1 above)
Comment (Frenzie) [reply assets/android.lua]: > I don't think using an unrecognized int will trigger an exception but I cannot test as the call is never triggered in the emulator and I currently don't have any working phone on any api below 28.

Well, you could test something even higher. ;-) I could take a look some other time what happens on Android 6/23 in any case, but let's not let that hold up this PR.

> Shall I add

Sure, let's just add those and call it a day for now.

Comment (pazos) [reply assets/android.lua]: ok, We don't need FLAG_IGNORE_VIEW_SETTING as we're a single view activity.

**Pull request merged**